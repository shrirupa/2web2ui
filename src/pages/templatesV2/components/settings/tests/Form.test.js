import { shallow } from 'enzyme';
import React from 'react';
import every from 'lodash/every';
import SettingsForm from '../Form';

jest.mock('../../DeleteTemplate');

describe('SettingsForm', () => {
  const subject = (props) => {
    const defaultProps = {
      domains: [
        { domain: 'test.com' },
        { domain: 'verified.com' }
      ],
      domainsLoading: false,
      hasSubaccounts: false,
      canViewSubaccount: false,
      isPublishedMode: false,
      handleSubmit: jest.fn((func) => func),
      showAlert: jest.fn()
    };

    return shallow(<SettingsForm {...defaultProps} {...props} />);
  };

  it('renders without subaccounts', () => {
    const wrapper = subject();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists('SubaccountSection')).toBe(false);
  });

  it('renders with subaccounts', () => {
    expect(subject({ hasSubaccounts: true, canViewSubaccount: true }).exists('SubaccountSection')).toBe(true);
  });

  it('renders with fields disabled in published mode', () => {
    const wrapper = subject({ hasSubaccounts: true, canViewSubaccount: true, isPublishedMode: true });
    const fieldProps = wrapper.find('Field').map((field) => field.prop('disabled'));
    expect(fieldProps.length).toEqual(10);
    expect(every(fieldProps)).toBe(true);
    expect(wrapper.find('SubaccountSection').prop('disabled')).toBe(true);
  });

  it('renders From Email help text with no verified domains', () => {
    const wrapper = subject({ domains: []});
    expect(wrapper.find('[name="content.from.email"]').prop('helpText'))
      .toEqual('You do not have any verified sending domains to use.');
  });

  it('renders From Email help text with no verified domains for subaccount', () => {
    const wrapper = subject({ subaccountId: 101, domains: []});
    expect(wrapper.find('[name="content.from.email"]').prop('helpText'))
      .toEqual('The selected subaccount does not have any verified sending domains.');
  });

  it('renders id field disabled', () => {
    expect(subject().find('[name="id"]').prop('disabled')).toBe(true);
  });

  describe('parseToggle', () => {
    const instance = subject().instance();
    it('converts value to boolean', () => {
      expect(instance.parseToggle('1')).toBe(true);
      expect(instance.parseToggle('')).toBe(false);
      expect(instance.parseToggle(true)).toBe(true);
      expect(instance.parseToggle(false)).toBe(false);
    });
  });

  describe('form submission', () => {
    it('renders submit button disabled', () => {
      const buttonSelector = 'Button[type="submit"]';

      expect(subject({
        submitting: false,
        valid: false,
        pristine: true
      }).find(buttonSelector).prop('disabled')).toBe(true);
      expect(subject({
        submitting: false,
        valid: false,
        pristine: false
      }).find(buttonSelector).prop('disabled')).toBe(true);
      expect(subject({
        submitting: true,
        valid: true,
        pristine: false
      }).find(buttonSelector).prop('disabled')).toBe(true);
    });

    it('updates settings', async () => {
      const mockUpdateDraft = jest.fn(() => Promise.resolve());
      const mockAlert = jest.fn();
      const wrapper = subject({
        valid: true,
        pristine: false,
        updateDraft: mockUpdateDraft,
        draft: { id: 'foo' },
        showAlert: mockAlert
      });
      await wrapper.find('form').simulate('submit');
      expect(mockUpdateDraft).toHaveBeenCalledWith({ id: 'foo' }, undefined);
      expect(mockAlert).toHaveBeenCalledWith({ type: 'success', message: 'Template settings updated.' });
    });

    it('updates settings with subaccount', () => {
      const mockUpdateDraft = jest.fn(() => Promise.resolve());
      const wrapper = subject({
        valid: true,
        pristine: false,
        updateDraft: mockUpdateDraft,
        draft: { id: 'foo' },
        subaccountId: 101
      });
      wrapper.find('form').simulate('submit');
      expect(mockUpdateDraft).toHaveBeenCalledWith({ id: 'foo' }, 101);
    });
  });

  describe('Delete Template', () => {
    it('renders delete template', () => {
      expect(subject().exists('DeleteTemplate')).toBe(true);
    });

    it('shows alert after delete and redirects', async () => {
      const mockAlert = jest.fn();
      const mockHistory = { push: jest.fn() };
      await subject({ showAlert: mockAlert, history: mockHistory }).find('DeleteTemplate').prop('afterDelete')();
      expect(mockAlert).toHaveBeenCalledWith({ type: 'success', message: 'Template deleted.' });
      expect(mockHistory.push).toHaveBeenCalledWith('/templatesv2');
    });
  });
});
