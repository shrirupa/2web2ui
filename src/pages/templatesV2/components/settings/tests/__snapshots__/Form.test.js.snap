// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm renders without subaccounts 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <Panel.Section>
      <Field
        component={[Function]}
        disabled={false}
        label="Template Name"
        name="name"
        validate={[Function]}
      />
      <Field
        component={[Function]}
        disabled={true}
        helpText="A Unique ID for your template, we'll fill this in for you."
        label="Template ID"
        name="id"
      />
    </Panel.Section>
    <Panel.Section>
      <Field
        component={[Function]}
        disabled={false}
        label="Subject"
        name="content.subject"
        validate={[Function]}
      />
      <Field
        component={[Function]}
        disabled={false}
        domains={
          Array [
            Object {
              "domain": "test.com",
            },
            Object {
              "domain": "verified.com",
            },
          ]
        }
        helpText={null}
        label="From Email"
        name="content.from.email"
        placeholder="example@email.com"
        validate={
          Array [
            [Function],
            [Function],
          ]
        }
      />
      <Field
        component={[Function]}
        disabled={false}
        helpText="A friendly from for your recipients."
        label="From Name"
        name="content.from.name"
      />
      <Field
        component={[Function]}
        disabled={false}
        helpText="An email address recipients can reply to."
        label="Reply To"
        name="content.reply_to"
        validate={[Function]}
      />
      <Field
        component={[Function]}
        disabled={false}
        helpText="Not visible to recipients."
        label="Description"
        name="description"
      />
    </Panel.Section>
    <Panel.Section>
      <Field
        component={[Function]}
        disabled={false}
        label="Track Opens"
        name="options.open_tracking"
        parse={[Function]}
        type="checkbox"
      />
      <Field
        component={[Function]}
        disabled={false}
        label="Track Clicks"
        name="options.click_tracking"
        parse={[Function]}
        type="checkbox"
      />
      <Field
        component={[Function]}
        disabled={false}
        helpText={
          <p
            className="HelpText"
          >
            Transactional messages are triggered by a userâ€™s actions on the website, like requesting a password reset, signing up, or making a purchase.
          </p>
        }
        label="Transactional"
        name="options.transactional"
        parse={[Function]}
        type="checkbox"
      />
    </Panel.Section>
    <Panel.Section>
      <Button
        disabled={true}
        primary={true}
        size="default"
        type="submit"
      >
        Update Settings
      </Button>
      <DeleteTemplate
        afterDelete={[Function]}
        className="DeleteButton"
      >
        Delete Template
      </DeleteTemplate>
    </Panel.Section>
  </form>
</Fragment>
`;
